### PR TITLE
Model interpretation

### PR DESCRIPTION
**Pull request by mllg**

### CODE REVIEW

**Comment by mllg** [comment]
Feature importance: I find it a bit odd to calculate 5% and 95% quantile error bars for 5 observations.

File: book/chapters/chapter12/model_interpretation.qmd [hunk 1]
```diff
@@ -7,26 +7,28 @@
...
+`counterfactuals` also uses the `R6` class system.
+In contrast to `iml` and `counterfactuals`, `DALEX` focuses on comparing multiple predictive models, usually of different types.
 We will only provide a brief overview of the methodology discussed below, we recommend @Molnar2022 as a comprehensive introductory book about IML.
 
-As a running example throughout this chapter, we will consider a gradient boosting machine (GBM) fit on half the features in the `"german_credit"` task:
+As a running example throughout this chapter, we will consider an gradient boosting machine (GBM) fit on half the features in the `"german_credit"` task.
+In practice, we would tune the hyperparameters of GBM as discussed in @sec-optimization and perform feature selection as discussed in @sec-feature-selection to select the most relevant features. 
+However, for the sake of simplicity, we opted to utilize an untuned GBM for the `"german_credit"` task, considering that it exhibited satisfactory performance even without fine-tuning, saving the reader valuable time and resources when running the example code below:
```
Comment (berndbischl) [inline book/chapters/chapter12/model_interpretation.qmd:31]: saving the reader valuable time and resources when running the example code below: --> remove

**Comment by berndbischl** [comment]
@RaphaelS1 pls check language. then merge and freeze

**Comment by berndbischl** [comment]
@mb706 has to add exercises

File: book/chapters/chapter12/model_interpretation.qmd [hunk 2]
```diff
@@ -150,16 +159,25 @@ effect = FeatureEffect$new(predictor, feature = "amount",
...
+
+The data used to train the black-box model may be very complex or limited, making it challenging to directly train a well-performing interpretable model on that data. 
+Having trained a black-box model that captures the relationships in the original data very well, enables us to generate new labeled data in specific regions of the input space with which we can augment the original data.
+<!-- During this 'surrogate data generation' process the predictions of the black-box model serve as labels. -->
+The augmented data can then be used to train an interpretable model that captures and explains the relationships learned by the black-box model (in specific regions) or to identify flaws or unexpected behavior.
+<!-- Surrogate data generation allows for the benefits of both the accuracy of the black-box model and the interpretability of the resulting interpretable model. -->
+<!-- The black-box model acts as a powerful predictor, providing labels for the artificially generated data, while the interpretable model can offer a better understanding in the targeted regions of the input space. -->
```
Comment (RaphaelS1) [inline book/chapters/chapter12/model_interpretation.qmd:180]: @mllg @giuseppec I'm sorry but I don't follow this at all, and I'm not entirely sure it's relevance to the chapter.
Why are we now talking about artifically generated data?

File: book/chapters/chapter12/model_interpretation.qmd [hunk 3]
```diff
@@ -168,35 +186,36 @@ This class extracts the decision rules created by the tree surrogate and the `$p
...
 
 ```{r iml-crosstable}
-pred_surrogate = treesurrogate$predict(credit_x, type = "class")
+pred_surrogate = tree_surrogate$predict(credit_x, type = "class")$.class
+pred_surrogate = factor(pred_surrogate, levels = c("good", "bad"))
 pred_gbm = lrn_gbm$predict_newdata(credit_x)$response
-prop.table(table(pred_gbm, pred_surrogate$.class))
-mean(pred_gbm == pred_surrogate$.class)
+confusion = mlr3measures::confusion_matrix(pred_surrogate, pred_gbm, positive = "good")
+print(confusion)
```
Comment (RaphaelS1) [inline book/chapters/chapter12/model_interpretation.qmd:200]:
```suggestion
confusion
```

File: book/chapters/chapter12/model_interpretation.qmd [hunk 4]
```diff
@@ -168,35 +186,36 @@ This class extracts the decision rules created by the tree surrogate and the `$p
...
 ```
```
Comment (RaphaelS1) [inline book/chapters/chapter12/model_interpretation.qmd:202]:
```suggestion
<!-- FIXME: RS CHECK PRINTOUT ABOVE -->
```

File: book/chapters/chapter12/model_interpretation.qmd [hunk 5]
```diff
@@ -209,44 +228,59 @@ Local surrogate models are constructed as follows:
...
 
 ```{r}
-c(global = global_predict[[1]], local = localsurrogate$predict()[[1]])
+ind_10nn = gower::gower_topn(Charlie, credit_x, n = 10)$index[, 1]
+Charlie_10nn = credit_x[ind_10nn, ]
+
+gbm_pred_10nn = predictor$predict(Charlie_10nn)[[1]]
+local_pred_10nn = local_surrogate$predict(Charlie_10nn)[[1]]
+mean(abs(gbm_pred_10nn - local_pred_10nn))
 ```
```
Comment (RaphaelS1) [inline book/chapters/chapter12/model_interpretation.qmd:271]:
```suggestion
<!-- FIXME: RS CHECK PRINTOUT -->
```

File: book/chapters/chapter12/model_interpretation.qmd [hunk 6]
```diff
@@ -296,7 +335,9 @@ For our `tsk("german_credit")` example, we might consider what changes in featur
 knitr::include_graphics("Figures/counterfactuals.png", dpi = 600)
 ```
 
-A simple counterfactual method is the `r index('What-If', aside = TRUE, lower = FALSE)` approach (@Wexler2019) where, for a given prediction to explain, the counterfactual is the closest data point in the dataset with the desired prediction -- so if the underlying dataset is changed then the counterfactual can be quite different.
+A simple counterfactual method is the `r index('What-If', aside = TRUE, lower = FALSE)` approach (@Wexler2019) where, for a given prediction to explain, the counterfactual is the closest data point in the dataset with the desired prediction.
+Obviously, if we do not restrict ourselves to a fixed dataset, many possible counterfactual data points can exist, as there are many ways to achieve a desired prediction.
+However, the approach by @Wexler2019 only produces a single counterfactual explanation, as many other early counterfactual methods (see @guidotti2022counterfactual for a comprehensive overview).
```
Comment (RaphaelS1) [inline book/chapters/chapter12/model_interpretation.qmd:340]: @mllg @giuseppec I think there's a typo here, I can't follow this sentence

File: book/chapters/chapter12/model_interpretation.qmd [hunk 7]
```diff
@@ -296,7 +335,9 @@ For our `tsk("german_credit")` example, we might consider what changes in featur
 knitr::include_graphics("Figures/counterfactuals.png", dpi = 600)
 ```
 
-A simple counterfactual method is the `r index('What-If', aside = TRUE, lower = FALSE)` approach (@Wexler2019) where, for a given prediction to explain, the counterfactual is the closest data point in the dataset with the desired prediction -- so if the underlying dataset is changed then the counterfactual can be quite different.
+A simple counterfactual method is the `r index('What-If', aside = TRUE, lower = FALSE)` approach (@Wexler2019) where, for a given prediction to explain, the counterfactual is the closest data point in the dataset with the desired prediction.
+Obviously, if we do not restrict ourselves to a fixed dataset, many possible counterfactual data points can exist, as there are many ways to achieve a desired prediction.
```
Comment (RaphaelS1) [inline book/chapters/chapter12/model_interpretation.qmd:339]:
```suggestion
If we do not restrict ourselves to a fixed dataset then many possible counterfactual data points can exist as there are many ways to achieve a desired prediction.
```

File: book/chapters/chapter12/model_interpretation.qmd [hunk 8]
```diff
@@ -306,14 +347,17 @@ The generation of counterfactuals is based on an optimization problem that aims
 3) Only require changes in a few features; and
 4) Originate from the same distribution as the observations in the given dataset.
 
-All four objectives are optimized simultaneously via a multi-objective optimization method.
-The returned set is a set of non-dominated and, therefore equally good, counterfactuals with respect to the four objectives (similarly to the `r index('Pareto front', lower = FALSE)` we saw in @sec-multi-metrics-tuning).
+In MOC, all four objectives are optimized simultaneously via a multi-objective optimization method.
+Conversely, several other counterfactual methods rely on single-objective optimization methods, where multiple objectives are combined into a single objective, e.g., using a weighted sum.
```
Comment (RaphaelS1) [inline book/chapters/chapter12/model_interpretation.qmd:351]:
```suggestion
Several other counterfactual methods rely on single-objective optimization methods, where multiple objectives are combined into a single objective, e.g., using a weighted sum. 
```

File: book/chapters/chapter12/model_interpretation.qmd [hunk 9]
```diff
@@ -264,20 +299,25 @@ The Shapley value of a feature is calculated by considering all possible subsets
...
 
 ```{r iml-006}
 #| fig-height: 3
 #| fig-cap: Shapley values for Charlie. The actual prediction (0.63) displays the prediction of the model for the observation we are interested in, the average prediction (0.71) displays the average prediction over the given test dataset. Each horizontal bar is the Shapley value (phi) for the given feature.
 #| fig-alt: 10 bar plots of Shapley values, one for each feature. x-axis says 'phi' and ranges from -0.1 to 0.05. The strongest positive contribution is from 'purpose=furniture/equipment; 'other_debtors=none' has the smallest value, and 'status=no checking account' is the strongest negative contribution.
 #| label: fig-iml-shapley
-shapley = Shapley$new(predictor, x.interest = Charlie)
+shapley = Shapley$new(predictor, x.interest = Charlie, sample.size = 1000)
 shapley$plot()
 ```
```
Comment (RaphaelS1) [inline book/chapters/chapter12/model_interpretation.qmd:315]:
```suggestion
<!-- fixme: rs check alt text -->
```

File: book/chapters/chapter12/model_interpretation.qmd (hunk 6 above)
Comment (berndbischl) [reply book/chapters/chapter12/model_interpretation.qmd:343]: rewrite like this:
A simple counterfactual method is the `r index('What-If', aside = TRUE, lower = FALSE)` approach (@Wexler2019) where, for a given prediction to explain, the counterfactual is the closest data point in the dataset with the desired prediction.
Usually, many possible counterfactual data points can exist. However, the approach by @Wexler2019, and many other early counterfactual methods for that matter (see @guidotti2022counterfactual for a comprehensive overview), only produce a single, somewhat arbitrary counterfactual explanation, which can be regarded as problematic when counterfactuals are used for insights or actions against the model.

File: book/chapters/chapter12/model_interpretation.qmd (hunk 2 above)
Comment (berndbischl) [reply book/chapters/chapter12/model_interpretation.qmd]: because artficially generated data is used to train the surrogate?

Comment (RaphaelS1) [reply book/chapters/chapter12/model_interpretation.qmd]: Okay I don't think that was said before or really here, how about:

> The data used to train the black-box model may be very complex or limited, making it challenging to directly train a well-performing interpretable model on that data. 
Instead, we can use the black-box model to generate new labeled data in specific regions of the input space with which we can augment the original data.
The augmented data can then be used to train an interpretable model that captures and explains the relationships learned by the black-box model (in specific regions) or to identify flaws or unexpected behavior.

File: book/chapters/chapter12/model_interpretation.qmd [hunk 10]
```diff
@@ -296,7 +338,9 @@ For our `tsk("german_credit")` example, we might consider what changes in featur
 knitr::include_graphics("Figures/counterfactuals.png", dpi = 600)
 ```
 
-A simple counterfactual method is the `r index('What-If', aside = TRUE, lower = FALSE)` approach (@Wexler2019) where, for a given prediction to explain, the counterfactual is the closest data point in the dataset with the desired prediction -- so if the underlying dataset is changed then the counterfactual can be quite different.
+A simple counterfactual method is the `r index('What-If', aside = TRUE, lower = FALSE)` approach (@Wexler2019) where, for a given prediction to explain, the counterfactual is the closest data point in the dataset with the desired prediction.
+If we do not restrict ourselves to a fixed dataset then many possible counterfactual data points can exist as there are many ways to achieve a desired prediction.
+However, the approach by @Wexler2019 only produces a single counterfactual explanation, as many other early counterfactual methods (see @guidotti2022counterfactual for a comprehensive overview).
```
Comment (RaphaelS1) [inline book/chapters/chapter12/model_interpretation.qmd:343]:
```suggestion
A simple counterfactual method is the `r index('What-If', aside = TRUE, lower = FALSE)` approach (@Wexler2019) where, for a given prediction to explain, the counterfactual is the closest data point in the dataset with the desired prediction.
Usually, many possible counterfactual data points can exist.
However, the approach by @Wexler2019, and several other early counterfactual methods (see @guidotti2022counterfactual for a comprehensive overview), only produce a single, somewhat arbitrary counterfactual explanation, which can be regarded as problematic when counterfactuals are used for insights or actions against the model.
```

File: book/chapters/chapter12/model_interpretation.qmd (hunk 2 above)
Comment (berndbischl) [reply book/chapters/chapter12/model_interpretation.qmd]: probably fine

**Review by RaphaelS1 — APPROVED** [review]

File: book/chapters/chapter12/model_interpretation.qmd [hunk 11]
```diff
@@ -150,16 +160,25 @@ effect = FeatureEffect$new(predictor, feature = "amount",
...
+
+The data used to train the black-box model may be very complex or limited, making it challenging to directly train a well-performing interpretable model on that data.
+Having trained a black-box model that captures the relationships in the original data very well, enables us to generate new labeled data in specific regions of the input space with which we can augment the original data.
+<!-- During this 'surrogate data generation' process the predictions of the black-box model serve as labels. -->
+The augmented data can then be used to train an interpretable model that captures and explains the relationships learned by the black-box model (in specific regions) or to identify flaws or unexpected behavior.
+<!-- Surrogate data generation allows for the benefits of both the accuracy of the black-box model and the interpretability of the resulting interpretable model. -->
+<!-- The black-box model acts as a powerful predictor, providing labels for the artificially generated data, while the interpretable model can offer a better understanding in the targeted regions of the input space. -->
```
Comment (RaphaelS1) [inline book/chapters/chapter12/model_interpretation.qmd:181]:
```suggestion
The data used to train the black-box model may be very complex or limited, making it challenging to directly train a well-performing interpretable model on that data.
Instead, we can use the black-box model to generate new labeled data in specific regions of the input space with which we can augment the original data.
The augmented data can then be used to train an interpretable model that captures and explains the relationships learned by the black-box model (in specific regions) or to identify flaws or unexpected behavior.
```

**Pull request merged**